### ISSUE TITLE
DB Refresh: Via objects API

### ISSUE THREAD

**Issue by joeflack4** [jhu-bids/TermHub#259]
## Overview
The current action to refresh the DB uses the datasets API. The primary disadvantages are that it takes a long time, and you have to do it or all tables at once.

Initially, this will only reflect new creates, not [updates and deletes](https://github.com/jhu-bids/TermHub/issues/261).

This will address the core concept set tables. Other kinds of objects will be handled in #341. Some tables will have to be updated using the datasets API, e.g. OMOP vocab tables (#340).

## Critical sub-tasks
- [x] 1. Logic: `enclave_wrangler.objects_api` functionality as needed
- [x] 2. Logic: Update DB w/ newly created objects
- [x] 3. CLI
- [x] 4. GitHub action (runs nightly, manually, and on API push)
- [x] 5. Logic: Update DB w/ newly updated objects
- [x] 6. Allow users to trigger updates (#376)

### Future sub-tasks
- #239
- #261 (depends on N3C internal infra team)
- Consider: Utilize `db_backup.sh` during refresh (will slow things down though)

### Related
- #340
- #341

### 6. Allow users to trigger updates
Add a button to the UI. If a user requests a refresh, send request w/ GitHub API to the GitHub action. But need some intermediary logic. If action is already running, don't want to interrupt it. Set variables in `manage` table, e.g. `refresh_status` (`done`, `running`), `last_refreshed`, `last_refresh_request`). After action completes, checks `last_refresh_request` and if that time is newer than the `last_refreshed` before the action began, it will begin again to do another refresh.

#### Tables to populate
| Table | Category | Logic |
|--------|--------|--------|
| <ul><li>[x] concept</li></ul> | OMOP: Vocab table | Nothing to be done here. @Sigfried However, currently doingas an OMOPConcept transform. I fetch new csets, and I add their concepts to this table if they're not there already. Potentially useful if we don't update our vocab tables as soon as the Enclave's are updated, but perhaps this is an unnecessary step I can remove. What do you think? |
| <ul><li>[x] relationship </li></ul> | OMOP: Vocab table | Nothing to be done at this time. See #340 | 
| <ul><li>[x] concept_relationship </li></ul> | OMOP: Vocab table | Nothing to be done at this time. See #340 | 
| <ul><li>[x] concept_ancestor </li></ul> | OMOP: Vocab table | Nothing to be done at this time. See #340 | 
| <ul><li>[x] concept_set_count_clamped </li></ul> | Enclave: derived from vocab | Nothing to be done at this time. See #340 However, I [asked Amin](https://cd2h.slack.com/archives/C047FEPM05U/p1685059896244699) how often and if there's a way to know when it's updated. |
| <ul><li>[x] deidentified_term_usage_by_domain_clamped </li></ul> | Enclave: derived from vocab | Nothing to be done at this time. See #340. See above comment on `concept_set_count_clamped`. |
| <ul><li>[x] omopconceptset </li></ul> | Enclave: derived from vocab? | Nothing to be done at this time? @Sigfried  This and omopconceptsetcontainer have ~4k rows. camelCase columns. I think we spoke before and you couldn't remember how we got these in the DB. I searched codebase but couldn't figure out. How are these used (I see that `all_csets` is created using `LEFT JOIN {{schema}}OMOPConceptSet` and `LEFT JOIN {{schema}}omopconceptsetcontainer`)? Do we need to keep these updated / is this within scope for this issue? | 
| <ul><li>[x] omopconceptsetcontainer </li></ul> | Enclave: derived from vocab? | Nothing to be done at this time? See above comment on `omopconceptset`. | 
| <ul><li>[x] concept_relationship_plus </li></ul> | TermHub: derived from vocab | Nothing to be done at this time. See #340  | 
| <ul><li>[x] concepts_with_counts </li></ul> | TermHub: derived from vocab | Nothing to be done at this time. See #340 | 
| <ul><li>[x] concept_set_json </li></ul> | TermHub: deprecated | @Sigfried Not sure if we should delete? Can see where it's used in the code [here](https://github.com/jhu-bids/TermHub/blob/db-refresh-objects/enclave_wrangler/objects_api.py#L549). | 
| <ul><li>[x] researcher</li></ul> | Enlave: ancillary object | Researcher transform. Will be done by #341. |
| <ul><li>[x] project</li></ul> | Enclave: ancillary object | ResearchProject transform. Haven't added to TermHub yet. Will be done by #341. |
| <ul><li>[x] review</li></ul> | Enclave: ancillary object | OMOPConceptSetReview transform. Haven't added to TermHub yet. Will be done by #341. |
| <ul><li>[x] code_sets</li></ul> | Enlave: core cset object | OMOPConceptSet transform |
| <ul><li>[x] concept_set_container</li></ul> | Enclave: core cset object | OMOPConceptSetContainer transform |
| <ul><li>[x] concept_set_version_item</li></ul> | Enclave: core cset object | OmopConceptSetVersionItem transform |
| <ul><li>[x] concept_set_members</li></ul> | Enclave: core cset object | First, query OMOPConceptSet. Then, get its all of its OMOPConcept members. | 
| <ul><li>[x] all_csets </li></ul> | Termhub: derived from core cset objects | Run modified DDL. | 
| <ul><li>[x] cset_members_items </li></ul> | Termhub: derived from core cset objects | Run modified DDL. | 
| <ul><li>[x] members_items_summary </li></ul> | Termhub: derived from core cset objects | Run modified DDL. | 
| <ul><li>[x] codeset_counts </li></ul> | Termhub: derived from core cset objects | Run modified DDL. | 
| <ul><li>[x] cset_member_items_plus </li></ul> | Termhub: derived from core cset objects | Run modified DDL. | 

##### Update process
I will likely go with 'b' for now because that's easiest. I can largely re-use Siggie's existing code in ddl.jinja.sql. Though in the future 'c' would be best.

(a) drop and recreate table

(b) create new table and rename
1. create new table (name append: _new)
2. rename old table (name append: _backup)
3. rename new table (name: original name)
4. drop old table

(c) update table
This may (not sure yet) require us first require us to modify our tables, either by (a) manually adding created_at and modified_at fields, or using PostgreSQL triggers (https://stackoverflow.com/questions/52426656/track-last-modification-timestamp-of-a-row-in-postgres).
We'd have to write separate logic to handle:
- creates
- updates
- deletes

##### Other considerations
I can't run all of `ddl.jinja.sql` every time an update occurs. I'll just want to run the parts of it I need. Ideally, running the necessary SQL for each table in separate, modular fashion.
I thought of these changes to `ddl.jinja.sql works` and went with 'a' (though I also did include `MODULE_NAME` as well).

(a) Split the whole of it into multiple files, and in `initalize()` we run them one at a time in the correct order, and during objects refresh, we run the parts we need in the correct order. Perhaps it would help to have numbers in the filenames to ensure correct order.

(b) Keep a single `ddl.jinja.sql` file, but make separate, `ddl-MODULE_NAME.jinja.sql` files. In `initialize()`, we would (i) instantiate the module templates first, and then we would change `ddl.jinja.sql` so that it takes each rendered template from 'i' and interpolates that into `ddl.jinja.sql`, e.g. `{{all_csets_rendered_template}}`.

**Comment by joeflack4** [comment]
Thanks @hugwuoke ! Yes, the only sub-task I'm working on now is #261, but no reason to keep this grouping issue open if that's the only one.

I also have some other follow-ups on my todo list that I need to do related to this. If they take longer than planned, I can open a separate issue.

@Sigfried I did tag you in some places there but I think you have more important stuff to do now. But I'd like you to be aware of the table with all the check boxes. It shows what my strategy was to update each given table.
